src/poetry/installation/executor.py

```diff
@@ -148,6 +150,19 @@ def execute(self, operations: list[Operation]) -> int:
                 if self._shutdown:
                     break
 
+                # It's not safe to clone the same git repository multiple times in parallel
+                if operation.package.source_type == "git" and not operation.skipped:
+                    # If this is the first operation cloning this repository,
+                    # allow it to be executed in parallel
+                    if operation.package.source_url not in git_repositories_to_clone:
+                        git_repositories_to_clone.add(operation.package.source_url)
+
+                    # If another operation is already cloning this repository,
+                    # it must be executed serially
+                    else:
+                        serial_operations.append(operation)
+                        continue
+
```

Comment:
This is a really cool fix! Two nits, though:
- it makes the loop logic much harder to read / understand, conceptually
- the serialization seems too strict: I suspect it's sufficient to serialize git ops on the same repo

I wonder if this could be made much simpler by bucketizing the git ops, and submitting them in a serialized fashion? e.g.
```python
serial_operations = []
git_serial_ops = collections.defaultdict(list)
```
Then right before `tasks.append`:
```python
if not operation.skipped and operation.package.source_type == "git":
    git_serial_ops[operation.package.source_url].append(operation)
    continue
```
Then right before the wait, submit the serialized ops:
```python
# git operations on the same repo url should execute serially
for ops in git_serial_ops.values():
    def serialize() -> None:
        for op in ops:
            self._execute_operation(op)
    tasks.append(self._executor.submit(serialize))
```

Reply:
I like this suggestion! I think the performance win is marginal, since the first operation on each repository is already done in parallel, **but** it turns two levels of conditionals into one, and to me this makes the loop easier to read.

Should I update this PR with the new proposed logic?

Reply:
I created a separate branch where I implemented this logic to make it easier to tell what it would look like: https://github.com/python-poetry/poetry/commit/8d15e5df192c13b8747e35d87c757e5bc6252be4

Tell me if you would like me to merge that branch into this one.

Reply:
Glad you liked it ! However I am just a poetry community member, and certainly not a maintainer. It's up to people such as @abn to decide which approach is preferable. I believe he's also the original author of the code in question.

Regarding the https://github.com/python-poetry/poetry/commit/8d15e5df192c13b8747e35d87c757e5bc6252be4 branch: nice, I like how you've refactored the operation logic. I don't think you need `functools.partial`: you can just define `_serialize` as a closure, which will capture and store `self` and `ops` in the function itself, like so:
```python
for ops in serial_git_operations.values():

  def _serialize() -> None:
    for op in ops:
      self._execute_operation(op)

  tasks.append(self._executor.submit(_serialize))
```

Reply:
I did write it like that at first, but was sternly warned by Ruff to not use a [loop variable in a function](https://docs.astral.sh/ruff/rules/function-uses-loop-variable/)!

Reply:
Nice gotcha! That just shows the value of linter tools to keep us on the python straight-and-narrow :)

How about this, would it be simpler than using `functools.partial` ?
```python
for ops in serial_git_operations.values():

  def _serialize(ops=ops) -> None:
    for op in ops:
      self._execute_operation(op)

  tasks.append(self._executor.submit(_serialize))
```

Reply:
> Tell me if you would like me to merge that branch into this one.

I took a quick glance and do not have any clear preference as to which approach is easier to understand. Since the second approach is functionally better in certain situations we should probably choose it.

Reply:
I switched to this solution.

However I did not remove the `functools.partial`, as the solutions are equivalent, and I personally have nothing against using `functools`. If someone feels strongly about this I will of course update it.

Reply:
Is `partial` really needed in this case? I.e. is your code:
```python
tasks.append(
    self._executor.submit(
        functools.partial(
            _serialize,
            repository_serial_operations=repository_git_operations,
        )
    )
)
```
equivalent to
```python
tasks.append(self._executor.submit(_serialize, repository_git_operations))
```
?